.github/workflows/ci.yml

```diff
@@ -60,7 +60,7 @@ jobs:
       - name: Run api tests
         run: cargo test -p bootloader_api
       - name: Run integration tests
-        run: cargo test
+        run: cargo test -- --test-threads=4
```

Comment:
Shouldn't it already be limited to the amount of cores by default?

Reply:
It seems that by default, rust gets the available "parallelism" from the OS: https://github.com/rust-lang/rust/blob/master/library/test/src/helpers/concurrency.rs

Reply:
Linux VM's get 2 cores, so `thread::available_parallelism()` should return 2.

Reply:
It doesn't seem to be, unfortunately. Another option would be to get the list of VM tests and run them with a matrix instead. I can write the code to generate the matrix for GH actions.

But looking at the logs, it appeared it was running at least 4 tests in parallel.

Reply:
And on top of that, oversubscription here wouldn't be terrible as long as it's limited. The tests spend a good amount of time doing IO, the CPU intense part is relatively quick.

Reply:
Perhaps the issue was simply a slow action runner.

I think a better solution here is to generate a matrix from the list of tests, and run them all separately. This has the added benefit of quickly seeing which test failed, and makes the logs easier to read.

This would run the tests in parallel on separate runners, which would reduce the time needed to run tests, and reduce the likelihood of a slow node impacting.

It will also easily allow more integration tests in the future, without risking a large slowdown like this.

Reply:
> Perhaps the issue was simply a slow action runner.

Maybe, but 20 minutes without any outputs should never happen, even on slow machines. So my guess is that some panic occured in the above case.

Regarding the general slowness of the Windows tests: Is QEMU really expected to be so slow on Windows? Maybe the issue is that we're running multiple threads at the same time. Could you try to update this PR to `--test-threads=1` to see whether this improves things?

One thing that we should definitely do is to increase the timeout for job, e.g. to `timeout-minutes: 60`.

> I think a better solution here is to generate a matrix from the list of tests, and run them all separately. This has the added benefit of quickly seeing which test failed, and makes the logs easier to read.

I'm not sure about this approach. It makes it easy to accidentally forget some tests (e.g. when a new test is added) and it spams the check list even more. Also, the number of free concurrent jobs is limited anyway, I think to 20 per _organization_. So running each test as a separate job will probably exceed this limit and lead to wait times, so we would not gain much.

Reply:
Ah, no, I mean automate, via building a matrix dynamically using the output of cargo test -- --list

Reply:
(we can also easily limit concurrency with the matrix too)

Reply:
> Regarding the general slowness of the Windows tests: Is QEMU really expected to be so slow on Windows? Maybe the issue is that we're running multiple threads at the same time. Could you try to update this PR to --test-threads=1 to see whether this improves things?
Added. Building as I write this.

> One thing that we should definitely do is to increase the timeout for job, e.g. to timeout-minutes: 60.
Also added.